File: middleware_py/policy.py
"""Policy engine for evaluating rules against tool calls."""

from dataclasses import dataclass, field
from typing import Any
import re
import time
import threading

from middleware_py.middleware_types import (
    Action,
    Condition,
    Decision,
    RateLimitConfig,
    ToolCall,
)


@dataclass
class Rule:
    """Individual policy rule."""

    name: str
    tools: list[str]
    action: Action
    conditions: list[Condition] = field(default_factory=list)
    max_rate: RateLimitConfig | None = None
    max_turns: int | None = None
    max_iterations: int | None = None
    redact_fields: list[str] = field(default_factory=list)

    def matches_tool(self, tool_name: str) -> bool:
        """Check if this rule applies to the given tool."""
        for pattern in self.tools:
            if pattern == "*":
                return True
            if pattern == tool_name:
                return True
        return False


@dataclass
class Policy:
    """Collection of policy rules."""

    rules: list[Rule] = field(default_factory=list)
    default_deny: bool = False

    def get_rules_for_tool(self, tool_name: str) -> list[Rule]:
        """Get all rules that apply to a given tool."""
        return [rule for rule in self.rules if rule.matches_tool(tool_name)]


class PolicyEvaluator:
    """Interface for evaluating policies."""

    def __init__(self, policy: Policy):
        self.policy = policy
        self._rate_limit_cache: dict[str, list[float]] = {}
        self._cache_lock = threading.Lock()
        self._turn_count: dict[str, int] = {}
        self._iteration_count: dict[str, int] = {}

    def evaluate(self, tool_call: ToolCall) -> Decision:
        """Evaluate policy for a tool call."""
        rules = self.policy.get_rules_for_tool(tool_call.tool_name)

        for rule in rules:
            if self._check_conditions(rule, tool_call):
                if not self._check_rate_limit(rule, tool_call):
                    return Decision(
                        action=Action.DENY,
                        rule_name=rule.name,
                        message=f"Rate limit exceeded for tool {tool_call.tool_name}",
                    )

                if not self._check_max_turns(rule, tool_call):
                    return Decision(
                        action=Action.DENY,
                        rule_name=rule.name,
                        message="Maximum turns exceeded",
                    )

                if not self._check_max_iterations(rule, tool_call):
                    return Decision(
                        action=Action.DENY,
                        rule_name=rule.name,
                        message="Maximum iterations exceeded",
                    )

                return Decision(
                    action=rule.action,
                    rule_name=rule.name,
                    redacted_fields=rule.redact_fields,
                )

        if self.policy.default_deny:
            return Decision(
                action=Action.DENY,
                message=f"No matching rule for tool {tool_call.tool_name}",
            )

        return Decision(action=Action.ALLOW)

    def _check_conditions(self, rule: Rule, tool_call: ToolCall) -> bool:
        """Check if all conditions in a rule match."""
        if not rule.conditions:
            return True

        for condition in rule.conditions:
            if not self._match_condition(condition, tool_call):
                return False
        return True

    def _match_condition(self, condition: Condition, tool_call: ToolCall) -> bool:
        """Match a single condition against the tool call."""
        value = self._resolve_field(condition.field, tool_call)
        return match_operator(condition.operator, value, condition.value)

    def _resolve_field(self, field: str, tool_call: ToolCall) -> Any:
        """Resolve a field path to its value."""
        if field.startswith("caller."):
            key = field[7:]
            return tool_call.caller_context.get(key)

        if field.startswith("args."):
            key = field[5:]
            return tool_call.args.get(key)

        if field.startswith("context."):
            key = field[8:]
            return tool_call.caller_context.metadata.get(key)

        return None

    def _check_rate_limit(self, rule: Rule, tool_call: ToolCall) -> bool:
        """Check rate limit for a rule."""
        if not rule.max_rate:
            return True

        session_id = tool_call.caller_context.session_id or "default"
        tool_name = tool_call.tool_name
        key = f"{session_id}:{tool_name}"

        with self._cache_lock:
            if key not in self._rate_limit_cache:
                self._rate_limit_cache[key] = []

            now = time.time()
            window = rule.max_rate.window

            self._rate_limit_cache[key] = [
                t for t in self._rate_limit_cache[key] if now - t < window
            ]

            if len(self._rate_limit_cache[key]) >= rule.max_rate.count:
                return False

            self._rate_limit_cache[key].append(now)
            return True

    def _check_max_turns(self, rule: Rule, tool_call: ToolCall) -> bool:
        """Check max turns limit."""
        if not rule.max_turns:
            return True

        session_id = tool_call.caller_context.session_id or "default"
        key = f"{session_id}:turns"

        count = self._turn_count.get(key, 0)
        if count >= rule.max_turns:
            return False

        self._turn_count[key] = count + 1
        return True

    def _check_max_iterations(self, rule: Rule, tool_call: ToolCall) -> bool:
        """Check max iterations limit."""
        if not rule.max_iterations:
            return True

        session_id = tool_call.caller_context.session_id or "default"
        tool_name = tool_call.tool_name
        key = f"{session_id}:{tool_name}:iterations"

        count = self._iteration_count.get(key, 0)
        if count >= rule.max_iterations:
            return False

        self._iteration_count[key] = count + 1
        return True


def match_operator(operator: str, field_value: Any, condition_value: Any) -> bool:
    """Match a field value against a condition using an operator."""
    if operator == "eq":
        return field_value == condition_value
    elif operator == "not_eq":
        return field_value != condition_value
    elif operator == "contains":
        return (
            field_value is not None
            and condition_value is not None
            and str(condition_value) in str(field_value)
        )
    elif operator == "not_contains":
        return (
            field_value is None
            or condition_value is None
            or str(condition_value) not in str(field_value)
        )
    elif operator == "matches":
        if field_value is None or condition_value is None:
            return False
        try:
            return bool(re.match(str(condition_value), str(field_value)))
        except re.error:
            return False
    elif operator == "not_matches":
        if field_value is None or condition_value is None:
            return True
        try:
            return not bool(re.match(str(condition_value), str(field_value)))
        except re.error:
            return False
    elif operator == "exists":
        return field_value is not None
    elif operator == "not_exists":
        return field_value is None
    elif operator == "not":
        return not field_value

    return False

File: middleware/policy.go
package middleware

import "time"

type PolicyEvaluator interface {
	Evaluate(toolName string, args map[string]any, caller CallerContext) Decision
}

type Policy struct {
	Rules       []Rule
	DefaultDeny bool
}

type Rule struct {
	Name         string
	Tools        []string
	Action       Action
	Conditions   []Condition
	MaxRate      *RateLimit
	RedactFields []string
}

type RateLimit struct {
	Count  int
	Window time.Duration
}

type Condition struct {
	Field    string
	Operator Operator
	Value    string
}

type Operator string

const (
	OperatorEq          Operator = "eq"
	OperatorNotEq       Operator = "not_eq"
	OperatorContains    Operator = "contains"
	OperatorNotContains Operator = "not_contains"
	OperatorMatches     Operator = "matches"
	OperatorNotMatches  Operator = "not_matches"
	OperatorExists      Operator = "exists"
	OperatorNotExists   Operator = "not_exists"
)

func (p *Policy) Evaluate(toolName string, args map[string]any, caller CallerContext) Decision {
	for _, rule := range p.Rules {
		if !rule.matchesTool(toolName) {
			continue
		}
		if !rule.evaluateConditions(args, caller) {
			continue
		}
		return Decision{
			Action:    rule.Action,
			Rule:      rule.Name,
			Reason:    "matched rule " + rule.Name,
			Timestamp: time.Now(),
		}
	}

	if p.DefaultDeny {
		return Decision{
			Action:    ActionDeny,
			Rule:      "default",
			Reason:    "no matching rules and default deny is enabled",
			Timestamp: time.Now(),
		}
	}

	return Decision{
		Action:    ActionAllow,
		Rule:      "default",
		Reason:    "no matching rules and default allow is enabled",
		Timestamp: time.Now(),
	}
}

func (r *Rule) matchesTool(toolName string) bool {
	for _, t := range r.Tools {
		if t == "*" || t == toolName {
			return true
		}
	}
	return false
}

func (r *Rule) evaluateConditions(args map[string]any, caller CallerContext) bool {
	if len(r.Conditions) == 0 {
		return true
	}
	for _, cond := range r.Conditions {
		if !cond.evaluate(args, caller) {
			return false
		}
	}
	return true
}

func (c *Condition) evaluate(args map[string]any, caller CallerContext) bool {
	var value string
	switch {
	case c.Field == "caller.role":
		value = caller.Role
	case c.Field == "caller.source":
		value = caller.Source
	case c.Field == "caller.trusted":
		if caller.Trusted {
			value = "true"
		} else {
			value = "false"
		}
	case c.Field == "caller.user_id":
		value = caller.UserID
	case c.Field == "caller.session_id":
		value = caller.SessionID
	case len(c.Field) > 5 && c.Field[:5] == "args.":
		argKey := c.Field[5:]
		if v, ok := args[argKey]; ok {
			if sv, ok := v.(string); ok {
				value = sv
			}
		}
	default:
		return false
	}

	switch c.Operator {
	case OperatorEq:
		return value == c.Value
	case OperatorNotEq:
		return value != c.Value
	case OperatorContains:
		return len(value) > 0 && len(c.Value) > 0 && contains(value, c.Value)
	case OperatorNotContains:
		return !contains(value, c.Value)
	case OperatorMatches:
		return matches(value, c.Value)
	case OperatorNotMatches:
		return !matches(value, c.Value)
	case OperatorExists:
		return value != ""
	case OperatorNotExists:
		return value == ""
	}
	return false
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 ||
		(len(s) > 0 && len(substr) > 0 && findSubstring(s, substr)))
}

func findSubstring(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

func matches(s, pattern string) bool {
	return simpleMatch(s, pattern)
}

func simpleMatch(s, pattern string) bool {
	if len(pattern) == 0 {
		return len(s) == 0
	}
	if len(s) == 0 {
		return false
	}
	return s == pattern
}
